tactic/back.lean

```diff
@@ -0,0 +1,112 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+
+import tactic.basic
+import tactic.ext
+
+namespace tactic
+
+meta def back_attribute : user_attribute := {
+  name := `back,
+  descr := "A lemma that should be applied to a goal whenever possible; use `back` to automatically `apply` all lemmas tagged `[back]`."
+}
+
+run_cmd attribute.register ``back_attribute
+
+/- TODO: replace this with an `!` parameter to `back`. -/
+meta def elim_attribute : user_attribute := {
+  name := `elim,
+  descr := "A lemma that should be applied during the `solve_by_elim` finishing step of `back`."
+}
+
+run_cmd attribute.register ``elim_attribute
+
+/-- Takes two sets of lemmas, 'finishing' lemmas and 'progress' lemmas.
+
+    Finishing lemmas should be applied only as part of a sequence of applications that close the goals.
+    Progress lemmas should be applied whenever possible, regardless of new hypotheses.
+
+    Finishing lemmas are applied in preference to progress lemmas.
+
+    `back` succeeds if at least one progress lemma was applied, or all goals are discharged.
+    -/
+
+-- TODO: Simon Hudon and Scott Morrison have discussed speeding this up, by indexing.
+meta def back (solve_by_elim_desc : string) (finishing_lemmas : list expr) :
+  list expr → bool → tactic (list string)
+| progress_lemmas progress :=
+  -- Note: the order here matters: we should priorise using assumptions,
+  -- over further applications of progress lemmas.
+  (do solve_by_elim_aux done (return (finishing_lemmas ++ progress_lemmas)) 3,
+      return [solve_by_elim_desc]) <|>
+  (do q ← (progress_lemmas.any_of $ λ e, apply_thorough e >>= (λ p, pure (e, p.2))),
+      r' ← back (progress_lemmas.erase q.1) tt,
+      return (q.2 ++ r')) <|>
+  guard progress >> pure []
+
+-- This is a close relative of `mk_assumption_set` in tactic/basic.lean.
+meta def mk_assumption_set (no_dflt : bool) (hs : list simp_arg_type) : tactic (list expr × list expr) :=
+do (hs, gex, hex, all_hyps) ← decode_simp_arg_list hs,
+   hs₁ ← hs.mmap i_to_expr_for_apply,
+   hs₂ ← attribute.get_instances `back,
+   hs₂ ← hs₂.mmap mk_const,
+   hs₃ ← attribute.get_instances `elim,
+   hs₃ ← hs₃.mmap mk_const,
+   let hs₃ := hs₃.filter $ λ h, expr.const_name h ∉ gex,
+   let pr := hs₁ ++ hs₂,
+   let pr := pr.filter $ λ h, expr.const_name h ∉ gex,
```

Comment:
@semorrison Can you explain what you're doing with the above two `filter` expressions? I have a feeling that this is not what you want.

Reply:
@cipher1024, I've renamed all the meaningless local variables, and restructured that code slightly (although I think not changed the intent).

The idea is that we are assembling two collections of lemmas the "progress lemmas" and the "finishing lemmas".

Progress lemmas are:
* everything explicitly named in the invocation of `back`
* any lemmas with attribute `back`

Finishing lemmas are:
* all local hypotheses which are not explicitly excluded (unless invoked with `back only ...`)
* any lemmas with attribute `elim`
* `congr_fun` and `congr_arg` (unless invoked with `back only ...`)